rosbag2_cpp/include/rosbag2_cpp/writers/sequential_writer.hpp

```diff
@@ -118,17 +123,55 @@ class ROSBAG2_CPP_PUBLIC SequentialWriter
 
   rosbag2_storage::BagMetadata metadata_;
 
+  // Used in invoking compression
+  rosbag2_cpp::CompressionMode compression_mode_;
+
   // Closes the current backed storage and opens the next bagfile.
   void split_bagfile();
 
   // Checks if the current recording bagfile needs to be split and rolled over to a new file.
   bool should_split_bagfile() const;
 
   // Prepares the metadata by setting initial values.
-  void init_metadata();
+  void init_metadata(const CompressionOptions & compression_options);
 
   // Record TopicInformation into metadata
   void finalize_metadata();
+
+  /**
+   * Check to make sure a valid bagfile split size is passed.
+   *
+   * \throws runtime_error If the size is too small (plugin specific).
+   */
+  void check_bagfile_size(const StorageOptions & storage_options);
+
+protected:
+  /**
+   * Initialize the compressor.
+   *
+   * \throws runtime_error If the compression implementation does not exist.
+   */
+  virtual void init_compression(const CompressionOptions & compression_options);
+
+  /**
+   * Checks if the compression by file option is specified and a compressor exists.
+   *
+   * If the above conditions are satisfied, compresses the most recent file and updates the
+   * metadata file paths.
+   *
+   * \return True if compression occurred, false otherwise.
+   */
+  virtual bool compress_file_and_update_metadata();
```

Comment:
Given the `if` condition check that's in there right now, returning `false` is definitely not an exception, it's a norm.

Reply:
I made it return a `bool` since we don't check if we should be compressing or not outside of this function. So if calling the function results in a compression we return true, false otherwise.

We could have a similar structure to bag file splitting where we have a `should_compress_file/message()` and check that before calling the compressor. I don't know if that's better from a design standpoint.

And in any case, `compress_uri()` throws when it's unable to compress.

Reply:
Okay, I realized throwing here is concerning, because the `~SequentialWriter()` destructor calls `reset()` which calls `compress_last_file()` which calls `compress_uri()`, which you're saying can potentially throw. This effectively means that the `~SequentialWriter()` destructor can throw.

Destructors that throw open up another whole can of worms https://isocpp.org/wiki/faq/exceptions#dtors-shouldnt-throw. There needs to be a try-catch somewhere, probably best in `~SequentialWriter()`.